src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc

```diff
@@ -177,14 +173,22 @@ class ScalarToTensorConstantMutator : public MixedModeMutator {
 };
 
 IRModule ScalarToTensorConstant(const IRModule& mod) {
-  auto mutator = ScalarToTensorConstantMutator(mod);
-  Function main_func = Downcast<Function>(mod->Lookup("main"));
-  auto new_main_body = mutator.VisitExpr(main_func->body);
-  if (!new_main_body.same_as(main_func->body)) {
-    auto main_var = mod->GetGlobalVar("main");
-    auto new_main_func = Function(main_func->params, new_main_body, main_func->ret_type,
-                                  main_func->type_params, main_func->attrs);
-    mod->Update(main_var, new_main_func);
+  for (auto gv : mod->GetGlobalVars()) {
+    Function func = Downcast<Function>(mod->Lookup(gv));
+
+    // only mutate CMSIS-NN external functions
+    auto compiler_name = func->GetAttr<String>(attr::kCompiler);
+    if (!compiler_name.defined() || compiler_name != "cmsis-nn") {
```

Comment:
There is no problem with this sort of check from top level visit. But maybe if it will have the same meaning if we could  move the checks inside function visits? Something similar to what is being done here: https://github.com/apache/tvm/blob/7127296c2b603e00344800f2d95aff5dc257f194/src/relay/backend/contrib/cmsisnn/extract_constants.cc#L60

Reply:
Although the pass won't change if we go the extract_constant() way, but your approach completely avoids changes to main(). Its a stricter check. So, let's go with that. Probably, something similar is needed in the extract_constant() then :smile: